contrib/pg_tde/documentation/docs/architecture.md

```diff
@@ -0,0 +1,403 @@
+
+There are workarounds for this, because operations that move the table data to a new file, such as `VACUUM FULL` or an `ALTER TABLE` that rewrites the file will create a new key for the new file, essentially rotating the internal key.
+This however means taking an exclusive lock on the table for the duration of the operation, which might not be desirable for huge tables.
+
+### Internal key storage
+
+Internal keys, and generally `pg_tde` metadata is kept in a single directory in `$PGDATA/pg_tde`. 
+In this directory each database has separate files, containing:
+
+* Encrypted internal keys and internal key mapping to tables
+* Information about the key providers
+
+There's also a special global section marked with the OID 607, which includes the global key providers / global internal keys.
+
+This is used by the WAL encryption, and can optionally be used by specific databases too, if global provider inheritance is enabled.
+
+### Key providers (principal key storage)
+
+Principal keys are stored on external providers.
+Currently pg_tde has 3 implementations:
+
+* A local file storage, intended for development and testing only
+* Hashicorp Vault
+* KMIP compatible servers
+
+In all cases, `pg_tde` requires a detailed configuration, including the address of the service and authentication information.
+With these details, all operations are done by `pg_tde` based on user operations:
+
+* When a new principal key is created, it will communicate with the service and upload a fresh key to it
+* When a principal key is required for decryption, it will try to get the key from the service
+
+### Sensitive key provider information
+
+Some of the keyring provider information (authentication details) is potentially sensitive, and it is not safe to store it together with the database in the `$PGDATA` directory, or even on the same server.
+
+For this purpose, `pg_tde` provides a mechanism where instead of directly specifying the parameter, users can specify an external service instead, from where it downloads the information.
+This way the configuration itself only contains the reference, not the actual authentication key or password.
+
+Currently only HTTP or external file references are supported, but other possible mechanisms can be added later, such as kubernetes secrets.
+
+## User interface
+
+### Setting up pg_tde
+
+To use `pg_tde`, users are required to:
+
+* Add pg_tde to the `shared_preload_libraries` in `postgresql.conf`, as this is required for the SMGR extensions
+* Execute `CREATE EXTENSION pg_tde` in the databases where they want to use encryption
+* Optionally, enable `pg_tde.wal_encrypt` in `postgresql.conf`
+* Optionally, disable `pg_tde.inherit_global_providers` in `postgresql.conf` (enabled by default)
+
+### Adding providers
+
+Keyring providers can be added to either the GLOBAL or to the database specific scope.
```

Comment:
Does it make sense to elaborate on what the GLOBAL scope is and what it is for?

Reply:
Yes as it is confusing

Reply:
But I'd ask a different question: do we need to have a User Interface section in the Architecture doc at all? It basically duplicates what we have in the whole doc set: how to set up pg_tde, what functions and params are there, etc.

Reply:
This is structured differently, as it was/is intended as a single document overview of everything.